### ISSUE TITLE
Plugin not working when a/i are remapped

### ISSUE THREAD

**Issue by ifsmirnov** [wellle/targets.vim#213]
Hi Christian!

I'm using Colemak layout and have some mappings in vim for it. In particular, I have `noremap k i`: I use `k` for entering Insert mode and as well in commands like `diw` (that is, I type `dkw` and expect it to behave like `diw`). Unfortunately, I couldn't get it fork with this plugin. I tried several things:

```
noremap k i
```
Standard commands like `diw` work, but targets' like `dia` do nothing (recall that I'm actually typing `dka`).

```
noremap k i
let g:targets_aiAI = 'akAK'
```
Same.

```
noremap k i
ounmap k
let g:targets_aiAI = 'akAK'
```
targets' commands start to work, but standard ones do not (that is, `diw` becomes `dkw` literally and erases one line above, then moving a word forward).

I tried several other combinations with mapping/unmapping `k` in different modes but didn't succeed.

Could you help me? I guess that the `targets_aiAI` setting was developed with similar case in mind but cannot grasp the solution.

**Comment by wellle** [comment]
@ifsmirnov: The problem is that (since https://github.com/wellle/targets.vim/pull/199) targets.vim uses a few expression mappings (for `i` etc.) instead of one mapping per text object, to reduce startup times. If you `noremap k i`, then when typing `k` an `i` is used, but because of the `nore` part this does not trigger our expression mapping. One thing I tried is use `map k i` instead, which does work.

But since you map `i` to `zvl` (no idea what that does), this means that if you type `k`, it gets mapped to `i`, which invokes targets.vim. In the `dkw` case targets doesn't find anything for `w`, so it returns `iw` back to Vim so Vim itself can handle the inner word text object. Because the mappings so far don't use `nore`, this `i` will trigger your `zvl` mapping, so `dkw` would still not work (but for another reason).

In general targets.vim currently relies on mapping `i`, `a`, `I` and `A` itself, so having your own mappings to and from those characters is going to lead to conflicts. One theoretical way could be to push this into targets.vim itself, so it would still own the `i` mappings, but could be set up to return `zvl` when it aborts. But I'm not sure if that would actually work. And it would probably also need some trickery similar https://github.com/wellle/targets.vim/pull/210 so that targets' expression mapping would behave as a `noremap` in case it aborts. But I'm not sure if I would want to go down that road.

You could try using version 5b1e0a5 which is not using expression mappings, together with
```vim
noremap k i
let g:targets_aiAI = 'akAK'
```

PS: There was a similar issue https://github.com/wellle/targets.vim/pull/32, not sure if there's anything helpful though.

**Comment by reaysawa** [comment]
I also use a abnormal keyboard setup (but did not try the suggested `onoremap` at #32) and what I found weird is that the Pair Text Objects worked fine, while the others didn't.

I ended up giving up on my noremapped/remap/map (doesn't make a difference) keys and setting other shortcuts instead, but I'll try the two-key onoremap approach and share my findings if it ends up well.

**Comment by wellle** [comment]
@reaysawa: Yes, the legacy mappings can't be extended like that.

@reaysawa @ifsmirnov: I just tried something to see if we can make this work and pushed it to branch [`213-mapped-aiAI`](https://github.com/wellle/targets.vim/compare/213-mapped-aiAI).

Then I put these lines into my vimrc to swap <kbd>k</kbd> and <kbd>i</kbd>:
```vim
inoremap k i
inoremap i k
nnoremap k i
nnoremap i k

let g:targets_aiAI        = 'akAI'
let g:targets_mapped_aiAI = 'aiAI'
```
(where `g:targets_mapped_aiAI` is a new setting that was just added on that branch)

I'm not quite sure if that is what you asked for. Can you try that and report back if that seems to work? Thank you!

**Comment by reaysawa** [comment]
@wellle sorry to get back kinda late on this. I have switched to that branch and it doesn't appear to have changed anything... Though I'm not very knowledgeable about this so I'll give more context on my situation.

I have remapped the arrow keys to spread across my home row

    nnoremap s h
    nnoremap e j
    nnoremap i k
    nnoremap o l

I want l/L to function as i/I for targets.vim, and o/s as directionals, so I put this in my config. 

    let g:targets_aiAI = 'alAL'
    let g:targets_nl = 'os'

**Not using legacy** when pressing `ca` I'm put into getchar() mode which gets my cursor indicator to the command line. I can use my o/s bindings normally after that, probably because I haven't mapped `a` to anything else in any mode. Works as it should.

**Not using legacy** pressing `cl` does not do the same. Instead I'm put into normal key sequence mode where I can build the target as I would with native text objects. Unlike `ca` I can't use my o/s bindings but the delimiters still work `}, ), ]...`.

![image](https://user-images.githubusercontent.com/17429390/50164348-36a16000-02c1-11e9-9c96-18b7f61eb887.png)

Following your branch

    let g:targets_aiAI        = 'alAL'
    let g:targets_mapped_aiAI = 'aiAI'

Has no effect as far as I can tell.

Lastly, I change my keys to

    let g:targets_aiAI = 'axAX'
    let g:targets_nl = 'os'

Then both `ca` and `cx` puts me in getchar() mode right away and all of the keys work correctly. As I mentioned before, it must be because both `A` and `X` aren't remapped to anything else in any mode.

---

**Using legacy** with

    call targets#legacy#addMappings('a', 'l', 'A', 'L', 'o', 's')

Both `ca` and `cl` put me in key sequence mode and all targets work normally. I have recorded a short video to show this. Take a look at the bottom right to see what I'm talking about.

[Legacy mappings](https://my.mixtape.moe/aelhac.mp4)

**Comment by wellle** [comment]
@reaysawa: I just tried what you described and it worked without any issues. Also, the case you're describing doesn't seem to be affected by mapping aiAI (so the branch isn't needed). As far as I'm aware the issue only comes up if you use a character in your aiAI settings which you're also trying to map globally. In your example you use `l` as `i`, but `l` is not mapped as far as I can tell (you only map `o` to `l`, but not `l` to anything else). Can you show me your full vimrc? I suspect you do map `l` somehow but didn't mention it above. Thank you!

**Comment by reaysawa** [comment]
MY BAD!! Since I use `I` for navigation now I've also remapped `L` to do insertion, as you have well guessed. I didn't understand the issue correctly, so sorry for wasting your time.

The other information about non-mapped keys `aA/xX` putting me in getchar() mode (non-legacy), while the remapped keys being unable do the same, still stands.

Could it be due to `nr2char(getchar())` not picking up the key correctly after conversion? I'll try to fiddle with the code later and see if I can fix it on my end.

---

" --- Editor configuration ---

set number
set relativenumber
set mouse=a
set wildignorecase
set expandtab
set clipboard^=unnamed,unnamedplus
set incsearch
set ignorecase
set smartcase
set noshowmode
set autoindent
set gdefault
set cursorline
set timeoutlen=1000
set ttimeoutlen=0
set shortmess+=cm

" For conceal markers.
if has('conceal')
  set conceallevel=2 concealcursor=niv
endif

set scrolloff=10
set tabstop=2
set shiftwidth=2

syntax enable
filetype on
filetype plugin on

let maplocalleader = " "
noremap <silent> <F5> :source ~/.config/nvim/init.vim<CR>
noremap <silent> <F6> :tabe ~/.config/nvim/init.vim<CR>
noremap <F7> :tabe ~/.config/
noremap <silent> ! q
noremap <silent> q :write<CR>
noremap <silent> Q :close<CR>

noremap s h
noremap e j
noremap i k
noremap o l
noremap h e
noremap l i
noremap L o
noremap H O
nnoremap m ^
vnoremap m ^
noremap , $
vnoremap , $<Left>
noremap ; ``

**Comment by wellle** [comment]
@reaysawa: Ok, I tried again and I think I got it working. Can you try removing these lines
```vim
noremap i k
noremap l i
```
and adding these?
```vim
inoremap i k
nnoremap i k
inoremap l i
nnoremap l i
let g:targets_mapped_aiAI = 'aiAI'
```
Explanation: Since you want to use `l` as `i` in `let g:targets_aiAI = 'alAL'`, it's important that you don't omap `l`. So instead of `noremap l i`, use `inoremap` and `nnoremap`. Now `clob` and similar should work. If you type `clp` (for `cip`, change in paragraph), the `let g:targets_mapped_aiAI = 'aiAI'` (added in the branch) returns `cip` back to Vim. But since you had `i` mapped to `k`, you need to use `inoremap` and `nnoremap` again for this to still work (don't have `i` omapped).

[Side note: If you wanted to actually omap `i` to say `x` (very unusual) you could use `x` instead of `i` in `targets_mapped_aiAI`, but lets ignore that for now]

Basically what's happening here is that we move the omapping part of these mappings into targets.vim. 

With all of those things combined, I believe it should work now. Please let me know 👍

**Comment by reaysawa** [comment]
It does work now! I hadn't paid attention that the noremap command also extended to object mode. From `:help map-modes`:

    Overview of which map command works in which mode. 
         COMMANDS                    MODES ~
    :map   :noremap  :unmap     Normal, Visual, Select, Operator-pending

Seems to be the same problem which @ifsmirnov was facing since he also uses `noremap k i` in the example.

Thanks very much for the help!

**Comment by wellle** [comment]
@reaysawa: Thanks for your help here! I just merged #220 to add this setting 👍
